crates/bevy_ecs/macros/src/lib.rs

```diff
@@ -423,6 +424,411 @@ pub fn derive_system_param(input: TokenStream) -> TokenStream {
+        } else if is_filter {
+            field_idents.push(field_ident.clone());
+            filter_field_idents.push(field_ident.clone());
+            query_types.push(query_type);
+            fetch_init_types.push(init_type);
+        } else {
+            field_idents.push(field_ident.clone());
+            non_filter_field_idents.push(field_ident.clone());
+            query_types.push(query_type);
+            fetch_init_types.push(init_type);
+        }
+        is_read_only = is_read_only && field_is_read_only;
+        readonly_types_to_assert.extend(field_readonly_types_to_assert.into_iter());
+    }
+
+    let read_only_impl = if is_read_only {
+        quote! {
+            /// SAFETY: each item in the struct is read only
+            unsafe impl #impl_generics #path::query::ReadOnlyFetch for #fetch_struct_name #ty_generics #where_clause {}
+
+            // Statically checks that the safety guarantee holds true indeed. We need this to make
+            // sure that we don't compile ReadOnlyFetch if our struct contains nested WorldQuery
+            // that don't implement it.
+            #[allow(dead_code)]
+            const _: () = {
+                fn assert_readonly<T: #path::query::ReadOnlyFetch>() {}
+
+                // We generate a readonly assertion for every type that isn't &T, &mut T, Option<&T> or Option<&mut T>
+                fn assert_all #impl_generics () #where_clause {
+                    #(assert_readonly::<<#readonly_types_to_assert as #path::query::WorldQuery>::Fetch>();)*
+                }
+            };
+        }
+    } else {
+        quote! {}
+    };
+
+    let tokens = TokenStream::from(quote! {
+        struct #fetch_struct_name #impl_generics #where_clause {
+            #(#field_idents: <#query_types as #path::query::WorldQuery>::Fetch,)*
+        }
+
+        struct #state_struct_name #impl_generics #where_clause {
+            #(#field_idents: <#query_types as #path::query::WorldQuery>::State,)*
+        }
+
+        impl #fetch_impl_generics #path::query::Fetch<#fetch_trait_punctuated_lifetimes> for #fetch_struct_name #fetch_ty_generics #where_clause {
+            type Item = #struct_name #ty_generics;
+            type State = #state_struct_name #fetch_ty_generics;
+
+            unsafe fn init(_world: &#path::world::World, state: &Self::State, _last_change_tick: u32, _change_tick: u32) -> Self {
+                #fetch_struct_name {
+                    #(#field_idents: <#fetch_init_types as #path::query::WorldQuery>::Fetch::init(_world, &state.#field_idents, _last_change_tick, _change_tick),)*
+                }
+            }
+
+            #[inline]
+            fn is_dense(&self) -> bool {
+                true #(&& self.#field_idents.is_dense())*
+            }
+
+            #[inline]
+            unsafe fn set_archetype(&mut self, _state: &Self::State, _archetype: &#path::archetype::Archetype, _tables: &#path::storage::Tables) {
```

Comment:
I would like to see `SAFETY` comments on these methods as well. They're straightforward enough, but macro code can be tough to read.